### ISSUE TITLE
The recognition result of classify/predict.py is very poor?

### ISSUE THREAD

**Issue by small-code-cat** [ultralytics/yolov5#9462]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I use classify/train.py to train the target classification, the model I have used yolov5s-cls.pt and efficient_v2_s, the first model was trained 50 times and the second one was trained 10 times. The loss has been decreasing and the accuracy is also good; however, the results of classify/predict.py are very poor
The prediction results are as follows.
<img width="1325" alt="image" src="https://user-images.githubusercontent.com/44658572/190884770-9b45aa4f-0811-4a0e-8788-e3addd7a8fe4.png">

**Comment by small-code-cat** [comment]
The dataset I use is minist. I hope someone can help me

**Comment by glenn-jocher** [comment]
@small-code-cat MNIST training runs correctly for me. I see 97.5% top 1 accuracy after two epochs with all default settings.

<img width="835" alt="Screenshot 2022-09-18 at 12 18 35" src="https://user-images.githubusercontent.com/26833433/190897259-d01534b0-8e23-4e62-9eea-721f3c8a5da0.png">

**Comment by small-code-cat** [comment]
Yes, my training results are as good as you, but look at my prediction results are very poor. I posted a screenshot of the prediction on it.
![BE3B5DB2-DA36-4199-8AEC-996E9454F32D](https://user-images.githubusercontent.com/44658572/190899408-e195e2af-1fd2-4e81-b3ea-649f3318e8f8.png)

**Comment by glenn-jocher** [comment]
@small-code-cat training is at imgsz 128, so you might want to predict at the same size.

**Comment by small-code-cat** [comment]
I don't quite understand. I didn't specify imgsz during training. Why do I need to specify when predicting? Isn't this automatically processed by the framework?

**Comment by glenn-jocher** [comment]
@small-code-cat see train.py and predict.py argparsers for details, i.e.:
https://github.com/ultralytics/yolov5/blob/4d50cd3469d75b18e99ce1e831ca024e3d25a2d7/classify/predict.py#L183-L206

**Comment by small-code-cat** [comment]
Thank you very much, it helped me a lot, it turns out that the imgsz of the two codes are not the same

**Comment by small-code-cat** [comment]
By the way, there is another question I want to ask. When using torch.hub.load to load the local classification model, it will report an error when going to predict images.
<img width="1512" alt="image" src="https://user-images.githubusercontent.com/44658572/190904083-73f91bc7-b228-4634-9ad2-397b9b46d450.png">

**Comment by glenn-jocher** [comment]
@small-code-cat ClassificationModel and SegmentationModel types are not yet fully supported by AutoShape, but you can still load them with PyTorch Hub and pass them torch tensors. Note you must apply your own pre and post processing:

<img width="1103" alt="Screenshot 2022-09-18 at 15 36 07" src="https://user-images.githubusercontent.com/26833433/190905013-723046b3-bec0-4ad7-bbe9-d2b52c0a3943.png">

**Comment by glenn-jocher** [comment]
@small-code-cat good news 😃! Your original issue may now be partially resolved ✅ in PR #9469. This PR aligns all default Classification task image sizes at 224 pixels. This should help avoid confusion in the future.

Note that many datasets like MNIST with very small image sizes (i.e. 28x28 pixels) can achieve good results at lower image resolutions like --imgsz 128, but ImageNet default is 224.

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by small-code-cat** [comment]
Thank you very much, this update is very useful, I hope yolov5 can do better and better, cheers!

**Comment by small-code-cat** [comment]
Thank you very much, this update is very useful, I hope yolov5 can do better and better, cheers!

**Comment by small-code-cat** [comment]
> @small-code-cat ClassificationModel and SegmentationModel types are not yet fully supported by AutoShape, but you can still load them with PyTorch Hub and pass them torch tensors. Note you must apply your own pre and post processing:
> 
> <img alt="Screenshot 2022-09-18 at 15 36 07" width="1103" src="https://user-images.githubusercontent.com/26833433/190905013-723046b3-bec0-4ad7-bbe9-d2b52c0a3943.png">

I tried it and it's running, thanks a lot!

**Comment by small-code-cat** [comment]
There is another question. After I train the model with gpu on Linux, loading the model on macos will report an error. What's the reason?
![7F1B9A8D-FFF4-47EF-ADBE-FA0D47DB6F45](https://user-images.githubusercontent.com/44658572/191001801-1267ae07-7590-4942-a884-b3a3afe679cb.jpeg)

**Comment by glenn-jocher** [comment]
@small-code-cat perhaps your paths are incorrect. I would just load as in the default example from ultralytics/yolov5

**Comment by small-code-cat** [comment]
I have tried this, and I will report this error.

**Comment by small-code-cat** [comment]
> @small-code-cat perhaps your paths are incorrect. I would just load as in the default example from ultralytics/yolov5

The problem has been solved, because I have a python file called utils in the same level directory, change the name and it is solved

**Comment by small-code-cat** [comment]
How do you remove the last few prediction layers when loading the model？ I want to use some of the layers to do kmeans clustering

**Comment by glenn-jocher** [comment]
@small-code-cat you can access the model's layers through the model's `model.model` attribute, which contains the actual YOLOv5 PyTorch model, and then modify it as needed. For example:

```python
import torch
from models.yolo import Model

model = torch.hub.load('ultralytics/yolov5', 'yolov5s')

# Access the actual YOLOv5 PyTorch model
yolo_model = model.model

# Remove the last few prediction layers
yolo_model.model[-1].pop()  # Remove the last layer from the YOLO layer list
yolo_model.model[-1].pop()  # Remove the second-to-last layer from the YOLO layer list

# Perform kmeans clustering using the modified model
# ...
```

Please note that modifying the model in this way will alter its behavior, so it's important to understand the implications of removing specific layers.

For more details on working with the YOLOv5 PyTorch model, you can refer to the [Ultralytics YOLOv5 documentation](https://docs.ultralytics.com/yolov5/).
